README.md
# sfcensus

## Installation

```
devtools::install_github("mvanhala/sfcensus")
```

## Overview

This is an R data package with US Census spatial data sets.

It contains several [`sf`](https://github.com/r-spatial/sf)-class data sets:

* `?state_sf`: 2019 boundaries for the 50 US states and DC, at 20M resolution

* `?county_sf`: 2019 boundaries for US counties, at 20M resolution

* `?place_sf`: 2019 boundaries for US places in the 50 states and DC, at 500K resolution

* `?zcta_sf`: 2010 boundaries for Zip Code Tabulation Areas, at 500K resolution

The `sf` objects are all [EPSG:4326](https://epsg.io/4326).

There are also a few other data sets in the package:

* `?states`: data frame of state names, abbreviations, and FIPS codes

* `?counties`: data frame of county names and FIPS codes

* `?zcta_county_rel`: data frame with the relationship between ZCTAs and counties. There is an
observation for each ZCTA-county intersection.




NEWS.md
# sfcensus 0.2.0

* Updated to 2020 data sets

* Added urban areas sf data frame

* Added a `NEWS.md` file to track changes to the package.


data-raw/boundaries.R

library(dplyr)
library(sf)

#### downloads --------------------------

# https://www.census.gov/geographies/reference-files/2020/demo/popest/2020-fips.html

curl::curl_download(
  "https://www2.census.gov/programs-surveys/popest/geographies/2020/all-geocodes-v2020.xlsx",
  "data-raw/files/all-geocodes-v2020.xlsx"
)


# https://www.census.gov/geographies/reference-files/time-series/geo/relationship-files.2020.html
# https://www2.census.gov/geo/docs/maps-data/data/rel2020/zcta520/tab20_zcta520_county20_natl.txt

curl::curl_download(
  "https://www2.census.gov/geo/docs/maps-data/data/rel2020/zcta520/tab20_zcta520_county20_natl.txt",
  "data-raw/files/tab20_zcta520_county20_natl.txt"
)


# https://www.census.gov/geographies/mapping-files/time-series/geo/cartographic-boundary.2020.html

# urban areas from TIGER/Line, cartographic boundary not available yet
# "Cartographic Boundary Files – May 2023"
# https://www.census.gov/programs-surveys/geography/guidance/geo-areas/urban-rural.html
# https://www.census.gov/geographies/mapping-files/time-series/geo/tiger-line-file.2020.html
# https://www2.census.gov/geo/tiger/TIGER2020/
# https://www2.census.gov/geo/tiger/TIGER2020/UAC/

curl::curl_download(
  "https://www2.census.gov/geo/tiger/GENZ2020/shp/cb_2020_us_state_20m.zip",
  "data-raw/files/cb_2020_us_state_20m.zip"
)

curl::curl_download(
  "https://www2.census.gov/geo/tiger/GENZ2020/shp/cb_2020_us_county_20m.zip",
  "data-raw/files/cb_2020_us_county_20m.zip"
)

curl::curl_download(
  "https://www2.census.gov/geo/tiger/GENZ2020/shp/cb_2020_us_zcta520_500k.zip",
  "data-raw/files/cb_2020_us_zcta520_500k.zip"
)

curl::curl_download(
  "https://www2.census.gov/geo/tiger/GENZ2020/shp/cb_2020_us_place_500k.zip",
  "data-raw/files/cb_2020_us_place_500k.zip"
)

curl::curl_download(
  "https://www2.census.gov/geo/tiger/TIGER2020/UAC/tl_2020_us_uac20.zip",
  "data-raw/files/tl_2020_us_uac20.zip"
)

curl::curl_download(
  "https://www2.census.gov/geo/tiger/TIGER2020/UAC/tl_2020_us_uac10.zip",
  "data-raw/files/tl_2020_us_uac10.zip"
)

unzip("data-raw/files/cb_2020_us_state_20m.zip", exdir = "data-raw/files/cb_2020_us_state_20m")
unzip("data-raw/files/cb_2020_us_county_20m.zip", exdir = "data-raw/files/cb_2020_us_county_20m")
unzip("data-raw/files/cb_2020_us_zcta520_500k.zip", exdir = "data-raw/files/cb_2020_us_zcta520_500k")
unzip("data-raw/files/cb_2020_us_place_500k.zip", exdir = "data-raw/files/cb_2020_us_place_500k")
unzip("data-raw/files/tl_2020_us_uac20.zip", exdir = "data-raw/files/tl_2020_us_uac20")

#### data creation ------------------------------

states <- readxl::read_excel("data-raw/files/all-geocodes-v2020.xlsx", skip = 3) |>
  filter(`Summary Level` == "040") |>
  select(
    state_fips = `State Code (FIPS)`,
    state_name = `Area Name (including legal/statistical area description)`
  ) |>
  inner_join(
    tibble(state_name = state.name, state_abb = state.abb) |>
      add_row(state_name = "District of Columbia", state_abb = "DC"),
    by = "state_name"
  ) |>
  arrange(state_fips)

counties <- read_sf("data-raw/files/cb_2020_us_county_20m") |>
  st_set_geometry(NULL) |>
  transmute(
    state_abb = STUSPS,
    state_fips = STATEFP,
    cty_fips = COUNTYFP,
    cty_name = NAME,
    cty_desc = na_if(stringr::str_replace_all(NAMELSAD, paste0("^", NAME, " ?"), ""), "")
  ) |>
  semi_join(states, by = "state_abb") |>
  arrange(state_fips, cty_fips)

zcta_county_rel <- readr::read_delim(
  "data-raw/files/tab20_zcta520_county20_natl.txt",
  delim = "|",
  col_types = readr::cols(
    "GEOID_ZCTA5_20" = "c",
    "GEOID_COUNTY_20" = "c",
    "AREALAND_ZCTA5_20" = "d",
    "AREALAND_PART" = "d",
    .default = "_"
  )
) |>
  filter(!is.na(GEOID_ZCTA5_20)) |>
  mutate(
    state_fips = stringr::str_sub(GEOID_COUNTY_20, end = 2),
    cty_fips = stringr::str_sub(GEOID_COUNTY_20, start = 3)
  ) |>
  inner_join(counties, by = c("state_fips", "cty_fips")) |>
  transmute(
    zcta = GEOID_ZCTA5_20,
    state_abb,
    state_fips,
    cty_fips,
    subset_area_land = AREALAND_PART,
    subset_area_land_pct = AREALAND_PART / AREALAND_ZCTA5_20
  ) |>
  arrange(zcta, state_fips, cty_fips)

zcta_state_rel <- zcta_county_rel |>
  group_by(zcta, state_abb) |>
  summarise(
    area_pct = sum(subset_area_land_pct),
    .groups = "drop"
  ) |>
  group_by(zcta) |>
  filter(row_number(desc(area_pct)) == 1) |>
  select(-area_pct) |>
  ungroup() |>
  arrange(zcta)

state_sf <- read_sf("data-raw/files/cb_2020_us_state_20m") |>
  select(state_fips = STATEFP, state_name =  NAME, state_abb = STUSPS) |>
  semi_join(states, by = "state_fips") |>
  arrange(state_fips) |>
  st_transform(4326)

county_sf <- read_sf("data-raw/files/cb_2020_us_county_20m") |>
  select(state_fips = STATEFP, cty_fips = COUNTYFP) |>
  inner_join(counties, by = c("state_fips", "cty_fips")) |>
  transmute(
    fips = paste0(state_fips, cty_fips),
    state_fips,
    cty_fips,
    state_abb,
    cty_name,
    cty_desc
  ) |>
  arrange(fips) |>
  st_transform(4326)

zcta_sf <- read_sf("data-raw/files/cb_2020_us_zcta520_500k") |>
  inner_join(zcta_state_rel, by = c("ZCTA5CE20" = "zcta")) |>
  select(zcta = ZCTA5CE20, state_abb) |>
  arrange(zcta) |>
  st_transform(4326)

place_sf <- read_sf("data-raw/files/cb_2020_us_place_500k") |>
  semi_join(states, by = c("STUSPS" = "state_abb")) |>
  transmute(
    state_abb = STUSPS,
    state_fips = STATEFP,
    place_fips = PLACEFP,
    place_name = NAME,
    place_desc = na_if(
      stringr::str_replace_all(
        NAMELSAD,
        paste0(
          "^",
          stringr::str_replace_all(NAME, c("\\(" = "\\\\\\(", "\\)" = "\\\\\\)")),
          " ?"
        ),
        ""),
      ""
    )
  ) |>
  arrange(state_fips, place_name, place_fips) |>
  st_transform(4326)

urban_area_sf <-read_sf("data-raw/files/tl_2020_us_uac20")|>
  transmute(
    code = UACE20,
    name = NAME20,
    states = stringr::str_extract_all(NAME20, paste(c(state.abb, "DC"), collapse = "|"))
  ) |>
  rowwise() |>
  filter(length(states) > 0) |>
  arrange(code) |>
  st_transform(4326)

##### save data ----------------------

usethis::use_data(state_sf, overwrite = TRUE)
usethis::use_data(county_sf, overwrite = TRUE)
usethis::use_data(zcta_sf, overwrite = TRUE)
usethis::use_data(place_sf, overwrite = TRUE)
usethis::use_data(urban_area_sf, overwrite = TRUE)

usethis::use_data(states, overwrite = TRUE)
usethis::use_data(counties, overwrite = TRUE)
usethis::use_data(zcta_county_rel, overwrite = TRUE)




R/data.R
#' @import tibble
#' @import sf
NULL

#' State names, abbreviations and FIPS codes
#'
#' A data frame containing the names, USPS abbreviations, and
#' FIPS codes of the 50 states and DC for the year 2020
#'
#' @format A data frame with 51 rows and 3 variables:
#' \describe{
#'   \item{state_fips}{FIPS code}
#'   \item{state_name}{name of the state}
#'   \item{state_abb}{two-letter USPS state abbreviation}
#' }
#' @source <https://www.census.gov/geographies/reference-files/2020/demo/popest/2020-fips.html>
"states"

#' County names and FIPS codes
#'
#' A data frame containing the names and FIPS codes of counties
#' in the fifty US states and DC for the year 2020
#'
#' @format A data frame with 3,143 rows and 5 variables:
#' \describe{
#'   \item{state_abb}{two-letter USPS state abbreviation}
#'   \item{state_fips}{two-digit state FIPS code}
#'   \item{cty_fips}{three-digit county FIPS code}
#'   \item{cty_name}{name of the county}
#'   \item{cty_desc}{county entity type}
#' }
#' @source <https://www.census.gov/geographies/reference-files/2020/demo/popest/2020-fips.html>
"counties"


#' ZCTA-county relationship
#'
#' A data frame with ZCTA to county relationships as of the 2020 Census.
#' There is a separate row for each ZCTA-county intersection record.
#' See <https://www2.census.gov/geo/pdfs/maps-data/data/rel2020/zcta520/explanation_tab20_zcta520_county20_natl.pdf>
#' for additional details.
#'
#' @format A data frame with 46,690 rows and 6 variables:
#' \describe{
#'   \item{zcta}{five-digit ZCTA code}
#'   \item{state_abb}{two-letter USPS state abbreviation}
#'   \item{state_fips}{two-digit state FIPS code}
#'   \item{cty_fips}{three-digit county FIPS code}
#'   \item{subset_area_land}{land area for relationship record, in square meters}
#'   \item{subset_area_land_pct}{percentage of total land area represented by record}
#' }
#' @source <https://www.census.gov/geographies/reference-files/time-series/geo/relationship-files.2020.html#list-tab-E7GGIXHXQT6PO7VYWR>
"zcta_county_rel"

#' sf data frame of US states
#'
#' An sf data frame of the fifty US states and DC.
#' It is the 2020 cartographic boundary file at resolution 20M.
#' The EPSG is 4326.
#'
#' @format An sf data frame with 51 rows and 4 variables:
#' \describe{
#'   \item{state_fips}{two-digit state FIPS code}
#'   \item{state_name}{name of the state}
#'   \item{state_abb}{two-digit USPS abbreviation}
#'   \item{geometry}{an sf geometry column with polygon data}
#' }
#' @source <https://www.census.gov/geographies/mapping-files/time-series/geo/cartographic-boundary.2020.html>
"state_sf"

#' sf data frame of US counties
#'
#' An sf data frame of counties in the fifty US states and DC.
#' It is the 2020 cartographic boundary file at resolution 20M.
#' The EPSG is 4326.
#'
#' @format An sf data frame with 3,143 rows and 7 variables:
#' \describe{
#'   \item{fips}{five-digit county FIPS code (i.e., concatenation of state and county FIPS codes)}
#'   \item{state_fips}{two-digit state FIPS code}
#'   \item{cty_fips}{three-digit county FIPS code}
#'   \item{state_abb}{two-digit USPS abbreviation}
#'   \item{cty_name}{name of the county}
#'   \item{cty_desc}{county entity type}
#'   \item{geometry}{an sf geometry column with polygon data}
#' }
#' @source <https://www.census.gov/geographies/mapping-files/time-series/geo/cartographic-boundary.2020.html>
"county_sf"

#' sf data frame of US Zip Code Tabulation Areas (ZCTA)
#'
#' An sf data frame of ZCTAs in the fifty US states and DC.
#' It is the 2020 cartographic boundary file at resolution 500K.
#' The EPSG is 4326.
#'
#' @format An sf data frame with 33,642 rows and 3 variables:
#' \describe{
#'   \item{zcta}{five-digit ZCTA code}
#'   \item{state_abb}{two-digit USPS abbreviation of state containing largest percentage of ZCTA population}
#'   \item{geometry}{an sf geometry column with polygon data}
#' }
#' @source <https://www.census.gov/geographies/mapping-files/time-series/geo/cartographic-boundary.2020.html>
"zcta_sf"


#' sf data frame of places
#'
#' An sf data frame of places in the fifty US states and DC.
#' It is the 2020 cartographic boundary file at resolution 500K.
#' The EPSG is 4326.
#'
#' @format An sf data frame with 31,617 rows and 6 variables:
#' \describe{
#'   \item{state_fips}{two-digit state FIPS code}
#'   \item{state_abb}{two-digit USPS abbreviation}
#'   \item{place_fips}{five-digit place FIPS code}
#'   \item{place_name}{place name}
#'   \item{place_desc}{place entity type}
#'   \item{geometry}{an sf geometry column with polygon data}
#' }
#' @source <https://www.census.gov/geographies/mapping-files/time-series/geo/cartographic-boundary.2020.html>
"place_sf"


#' sf data frame of urban areas
#'
#' An sf data frame of urban areas in the fifty US states and DC.
#' It is the 2020 TIGER/Line file.
#' The EPSG is 4326.
#'
#' @format An sf data frame with 2,613 rows and 4 variables:
#' \describe{
#'   \item{code}{five-digit urban area code}
#'   \item{name}{urban area name}
#'   \item{states}{vector of states of the urban area}
#'   \item{geometry}{an sf geometry column with polygon data}
#' }
#' @source <https://www.census.gov/geographies/mapping-files/time-series/geo/cartographic-boundary.2020.html>
"urban_area_sf"


